x/ecocredit/server/basket/put.go

```diff
@@ -2,11 +2,181 @@ package basket
-	//TODO implement me
-	panic("implement me")
+func (k Keeper) Put(ctx context.Context, req *baskettypes.MsgPut) (*baskettypes.MsgPutResponse, error) {
+	ownerAddr, _ := sdk.AccAddressFromBech32(req.Owner)
+
+	// get the basket
+	basket, err := k.stateStore.BasketStore().GetByBasketDenom(ctx, req.BasketDenom)
+	if err != nil {
+		return nil, err
+	}
+
+	// keep track of the total amount of tokens to give to the depositor
+	amountReceived := sdk.NewInt(0)
+	for _, credit := range req.Credits {
+		// get credit batch info
+		res, err := k.ecocreditKeeper.BatchInfo(ctx, &ecocredit.QueryBatchInfoRequest{BatchDenom: credit.BatchDenom})
+		if err != nil {
+			return nil, err
+		}
+		batchInfo := res.Info
+
+		// validate that the credit batch adheres to the basket's specifications
+		if err := k.validateCredit(ctx, basket, batchInfo); err != nil {
+			return nil, err
+		}
+		// get the amount of credits in dec
+		amt, err := regenmath.NewPositiveFixedDecFromString(credit.Amount, basket.Exponent)
+		if err != nil {
+			return nil, err
+		}
+		// update the user and basket balances
+		if err = k.updateBalances(ctx, ownerAddr, amt, basket, batchInfo); err != nil {
+			return nil, err
+		}
+		// get the amount of basket tokens to give to the depositor
+		tokens, err := calculateTokens(amt, basket.Exponent, basket.BasketDenom)
+		if err != nil {
+			return nil, err
+		}
+		// update the total amount received so far
+		amountReceived = amountReceived.Add(tokens[0].Amount)
+		// mint and send tokens to depositor
+		sdkCtx := sdk.UnwrapSDKContext(ctx)
+		if err = k.bankKeeper.MintCoins(sdkCtx, ecocredit.ModuleName, tokens); err != nil {
+			return nil, err
+		}
+		if err = k.bankKeeper.SendCoinsFromModuleToAccount(sdkCtx, ecocredit.ModuleName, ownerAddr, tokens); err != nil {
+			return nil, err
+		}
+	}
+	return &baskettypes.MsgPutResponse{AmountReceived: amountReceived.String()}, nil
+}
+
+// validateCredit checks that a credit adheres to the specifications of a basket. Specifically, it checks:
+//  - batch's start time is within the basket's specified time window or min start date
+//  - class is in the basket's allowed class store
+//  - type matches the baskets specified credit type.
+func (k Keeper) validateCredit(ctx context.Context, basket *basketv1.Basket, batchInfo *ecocredit.BatchInfo) error {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	blockTime := sdkCtx.BlockTime()
+
+	if basket.DateCriteria != nil && basket.DateCriteria.Sum != nil {
+		// check time window match
+		var minStartDate time.Time
+		switch basket.DateCriteria.Sum.(type) {
+		case *basketv1.DateCriteria_MinStartDate:
+			min := basket.DateCriteria.Sum.(*basketv1.DateCriteria_MinStartDate).MinStartDate
+			minStartDate = min.AsTime()
+		case *basketv1.DateCriteria_StartDateWindow:
+			window := basket.DateCriteria.Sum.(*basketv1.DateCriteria_StartDateWindow).StartDateWindow.AsDuration()
+			minStartDate = blockTime.Add(window)
```

Comment:
```suggestion
			minStartDate = blockTime.Sub(window)
```

Reply:
Sub is an operation for time.Time, it doesn't accept duration